include self link in schema

NobodysNightmare opened the pull request:
Fixes test failure of https://community.openproject.org/work_packages/18570
## Notes
- self link will only be displayed outside of a form, inside a form there is additional information that is not reachable via a link
- `_type` indicating `Schema` is still missing... this goes down to a flaw in the specification, because we want two things:
  - indicate that there is a field called `_type` on the thing described by the schema
  - indicate that the schema has a `_type` of `Schema`

With the current schema layout it is not possible to do both, because that would result in a duplicate key called `_type`. We therefore only indicate the former. At a later point the specification on this needs to be revised.

I would really apprechiate if you could **merge #2550 _before_ merging this PR**. This would move any conflicts into the smaller PR.

NobodysNightmare committed:
include self link in schema

- only if the schema is not embedded in a form

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-fix-self-link-schema.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-fix-self-link-schema.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
fix documentation example

[ci skip]

NobodysNightmare committed:
style

ulferts commented:
@NobodysNightmare needs rebasing, sorry.

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/dev' into fix/self_link_schema

Conflicts:
	lib/api/v3/work_packages/form/form_representer.rb
	lib/api/v3/work_packages/schema/work_package_schema_representer.rb
	spec/lib/api/v3/work_packages/work_package_schema_representer_spec.rb

ulferts commented:
This is broken and needs rebasing @NobodysNightmare

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/dev' into fix/self_link_schema

Conflicts:
	lib/api/decorators/schema.rb

NobodysNightmare committed:
fux failing specs

NobodysNightmare committed:
some refactoring around the WP schema

- schema does not decide anymore whether to include allowed values (solely decided by "form_embedded" of representer)
- schema will answer as many "assignable_*" methods as possible
- work package does not need those methods anymore (except few cases)

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -1608,7 +1608,6 @@ Note that regular expressions used in the API follow the rules of [Ruby Regular
     + Body
 
             {
-                "_type": "Schema",
```
Shouldn't we write what you wrote as a preamble to this PR into the documentation as well?

NobodysNightmare replied:
I'd like to update the documentation as soon as we have the real solution to be honest...

As nobody seems to be interested in requiring the `_type` for `PATCH` requests I'd say lets change that now:
1. I reintroduce the `"_type": "Schema"`
2. I remove `_type` attributes from schema listings
3. Specification is updated to state that the `_type` is not required for `PATCH`ing

Any objections?

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema.rb:
```diff
@@ -63,19 +59,19 @@ def assignable_statuses_for(user)
 
           def assignable_priorities
-            @work_package.assignable_priorities if defines_assignable_values?
+            IssuePriority.where(active: true)
```
Note to self and whoever reads this. It would be better to have an `active` scope on IssuePriority to avoid AR methods here. I am aware, that the code was just moved, though.

NobodysNightmare replied:
How hard is it to add this scope?

Should I just do it?

NobodysNightmare replied:
Is it more than just `scope :active, where(active: true)`?

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema.rb:
```diff
@@ -63,19 +59,19 @@ def assignable_statuses_for(user)
 
           def assignable_versions
-            @work_package.assignable_versions if defines_assignable_values?
+            @work_package.assignable_versions unless @work_package.nil?
```
For patterns like this you can use:

`@work_package.try(:assignable_versions)`

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema.rb:
```diff
@@ -63,19 +59,19 @@ def assignable_statuses_for(user)
 
           def available_custom_fields
-            @project.all_work_package_custom_fields & @type.custom_fields.all
+            project.all_work_package_custom_fields & type.custom_fields.all
```
@NobodysNightmare will this conflict with what you are doing in the other PR right now? Any preference of merging order?

NobodysNightmare replied:
Git will tell us if it is conflicting :)

Merge whatever is ready first.

ulferts commented:
@NobodysNightmare does what it says and I could spot no flaw.

I'd still like to discuss whether it is correct to have the schema **embedded** inside the work package form, thereby suggesting that it is a resource it it's own right, and then not having a self link to said resource. The standard is not my strong suit but I think that we violate it here and I am wondering whether we can simply move the schema to anything other than the embedded section. Will that make us  good again?

NobodysNightmare commented:
@ulferts regarding standards conformity:

It might sound crazy, but I am also not pedantic enough to see where we are violating which standard. However, I will try my best.

What I have seen so far in the wild, and what we are also doing is that it is possible to have stuff embedded, that is not linkable. e.g. looking at our collections they will always have `_embedded/elements`, though there is no link for those elements except the collection itself (which only carries those elements as embedded).

Looking at the form we already have the `payload`, which is also **only** embedded. It has no self link and no self link would be valid, because the payload is an imaginary object that only becomes real (as in "has a link") as soon as you call the `PATCH` action.

**tl; dr:** I don't know what the HAL-police will say, but looking at the rest of _our_ API the embedded schema seems to be fitting (the rest is at least on the same degree of wrong-ish-ness).

NobodysNightmare committed:
use "try"

NobodysNightmare committed:
extract a scope for active priorities

NobodysNightmare commented:
I changed the specification to not require specifying a `_type` in `PATCH` anymore. Therefore I could make the schema more compliant with all other resources.

old schema:

``` JSON
{
  "_type": {
    "type": "MetaType",
    "...": "..."
  },
  "subject": {
    "type": "String",
    "...": "..."
  },
}
```

new schema:

``` JSON
{
  "_type": "Schema",
  "subject": {
    "type": "String",
    "...": "..."
  },
}
```

This means that a schema resource now should look more similar to all the other resources out there.

@0xF013 I don't know if this is a thing for you, but you might want to know.

NobodysNightmare committed:
make _type in schema indicate the schemas type

instead of indicating that the described resource has something
called "_type"

NobodysNightmare commented:
Go home @houndci, you are drunk!

NobodysNightmare committed:
use callable to define scope

- Rails 4 compatible
- MIGHT fix an issue on travis

NobodysNightmare committed:
remove dead property

ulferts commented:
@NobodysNightmare needs rebasing, too.

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/dev' into fix/self_link_schema

Conflicts:
	app/models/work_package.rb
	lib/api/v3/work_packages/schema/work_package_schema.rb
	spec/lib/api/v3/work_packages/schema/work_package_schema_spec.rb
	spec/lib/api/v3/work_packages/work_package_schema_representer_spec.rb
